### PR TITLE
Rescale image data if outside `float32` precision 

### LINKED ISSUE #5847: Closes widgets via X isn't registered as unticking in the plugin menu
## 🐛 Bug

Refers to this issue discovered using a [napari-matplotlib plug in](https://github.com/matplotlib/napari-matplotlib/issues/124), but happens with other plugins

## To Reproduce

1. Plugins menu --> select widget e.g. Histogram (tick appears next to Histogram and widget appears)
2. Close the Histogram widget using the X in the top left corner of the widget so it disappears
3. Plugins menu Histogram widget still has a tick next to it and clicking Histogram again makes the tick disappear and the widget appear so the functionality is swapped

## Environment

 - Please copy and paste the information at napari info option in help menubar here:

 - Any other relevant information:

### LINKED ISSUE #6533: Canvas goes black for very small or large numbers in image
### 🐛 Bug Report

When dealing with Nd arrays containing slices with very small or large values, sometimes the image disappears (goes black, under the gray colormap) on the canvas. This seems to be caused by an overflow error.  

### 💡 Steps to Reproduce

```
# imports
import numpy as np
import napari

# create image array
arr = np.linspace(1,9,5*5,dtype=np.float64).reshape((5,5))

# open the image
viewer, image_layer = napari.imshow(arr)
```

1. Click on 'continuous' auto-contrast.

```
viewer.layers[0].data = viewer.layers[0].data*1E37
```

2. The image is still visible in on the canvas.

```
viewer.layers[0].data = viewer.layers[0].data*10
```
3. The canvas has gone black! However, the image is still visible in the preview of the layer in the layer list, and hovering the mouse over the canvas still shows the values under the cursor in the Status Bar. Adjusting the contrast limits slider still varies the contrast in the preview correctly. I get the following error message:

```
C:\Users\...\napari\napari\_vispy\utils\gl.py:120: RuntimeWarning: overflow encountered in cast
  return data.astype(dtype_)
>>> C:\Users\...\Anaconda3\envs\napari-env\lib\site-packages\vispy\gloo\program.py:373: RuntimeWarning: overflow encountered in cast
  data = np.array(data, dtype=dtype).ravel()
```
Keep going. 
```
viewer.layers[0].data = viewer.layers[0].data/10
```
5. The image is back!
```
viewer.layers[0].data = viewer.layers[0].data/1E-75
```
6. Still visible...
```
viewer.layers[0].data = viewer.layers[0].data/10
```
7. And gone.

### 💡 Expected Behavior

The image should be visible on the canvas for any range of data values allowed by the `dtype`.

### 🌎 Environment

napari: 0.5.0a2.dev486+g4d60a7ce
Platform: Windows-10-10.0.19045-SP0
Python: 3.9.18 | packaged by conda-forge | (main, Aug 30 2023, 03:40:31) [MSC v.1929 64 bit (AMD64)]
Qt: 5.15.2
PyQt5: 5.15.10
NumPy: 1.26.2
SciPy: 1.11.4
Dask: 2023.12.0
VisPy: 0.14.1
magicgui: 0.8.0
superqt: 0.6.1
in-n-out: 0.1.9
app-model: 0.2.2
npe2: 0.7.3

OpenGL:
  - GL version:  4.6.0 - Build 27.20.100.9565
  - MAX_TEXTURE_SIZE: 16384

Screens:
  - screen 1: resolution 1920x1200, scale 1.0
  - screen 2: resolution 2000x1125, scale 1.0

Settings path:
  - C:\Users\...\AppData\Local\napari\napari-env_4624f08893c59545bcb98518a8c42a23d61d4ece\settings.yaml

### 💡 Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Czaki**
# References and relevant issues

closes #6533 

# Description

Add check if image data could be visualized using vispy and if not, then rescale data.

### CODE REVIEW

**Comment by jnahlers** [comment]
Thanks @Czaki, this method looks like it handles the toy example in the 'Steps to Reproduce' of #6533. However, I am still having trouble with my actual images. An example is [this image](https://drive.google.com/file/d/1DKs1A2f-L73lTsHOoH1w3OEPB3Z_izl2/view?usp=drive_link).

**Comment by Czaki** [comment]
@jnahlers Your case is different from described in the original issue. In this case, is small contrast limits for big values. But it looks like I could easily fix the code to work with your data.

**Comment by jnahlers** [comment]
It looks like there is still some kind of overflow happening? Check out [this array](https://drive.google.com/file/d/1LSuDoQDZ3kw0H__hvGIdp-Di7DGR63Ch/view?usp=drive_link), specifically  around index 50 (i.e. [50,:,:]).

**Comment by Czaki** [comment]
Now It should be better. I forgot to cast infinity to float32.

**Comment by Czaki** [comment]
NumPy 1.22.0 and 1.22.1 somehow overrides python logger and breaks the tests.

File: napari/layers/image/image.py [hunk 1]
```diff
@@ -455,7 +456,7 @@ def custom_interpolation_kernel_2d(self, value):
         self._custom_interpolation_kernel_2d = np.array(value, np.float32)
         self.events.custom_interpolation_kernel_2d()
 
-    def _raw_to_displayed(self, raw):
+    def _raw_to_displayed(self, raw: np.ndarray) -> np.ndarray:
         """Determine displayed image from raw image.
 
         For normal image layers, just return the actual image.
```
Comment (melonora) [inline napari/layers/image/image.py:462]: Can this be removed? I see you raise `NotImplementedError`, but you do have a function with the same name at line 1195.

File: napari/_vispy/layers/image.py [hunk 2]
```diff
@@ -326,7 +327,8 @@ def _update_mip_minip_cutoff(self) -> None:
                 self.node.minip_cutoff = None
 
     def _on_contrast_limits_change(self) -> None:
-        self.node.clim = self.layer.contrast_limits
+        lim_tup = _coerce_contrast_limits(self.layer.contrast_limits)
```
Comment (melonora) [inline napari/_vispy/layers/image.py]: I prefer complete variable names. I am not certain what you mean with `lim_tup`.

**Review by melonora — COMMENTED** [review]
will look further later

File: napari/layers/image/image.py (hunk 1 above)
Comment (Czaki) [reply napari/layers/image/image.py:462]: Yes, but this functions belong to `_ImageBase` and this from line 1195 belongs to class `Image`

File: napari/_vispy/layers/image.py (hunk 2 above)
Comment (Czaki) [reply napari/_vispy/layers/image.py]: removed

File: napari/layers/image/image.py (hunk 1 above)
Comment (melonora) [reply napari/layers/image/image.py:462]: ah ok I see.

Comment (Czaki) [reply napari/layers/image/image.py:462]: I think that _ImageBase should be moved to separate file, but it should be separate PR. And we should Identify first all PR that touches this code.

File: napari/utils/colormaps/colormap_utils.py [hunk 3]
```diff
@@ -906,3 +906,52 @@ def display_name_to_name(display_name):
...
+    ):
+        return scale_up(contrast_limits)
+
+    return CoercedContrastLimits(contrast_limits, 0, 1)
+
+
+def scale_down(contrast_limits: Tuple[float, float]):
+    """Scale down contrast limits to be in the float32 range."""
+    vispy_max = float(np.finfo(np.float32).max / 8)
+    # Using 8 as divisor comes from experiments.
```
Comment (psobolewskiPhD) [inline napari/utils/colormaps/colormap_utils.py]:
```suggestion
    # Using 8 as divisor comes from experiments.
```
Maybe put this comment above, where the magic 8 is first used.

File: napari/utils/colormaps/colormap_utils.py [hunk 4]
```diff
@@ -906,3 +906,52 @@ def display_name_to_name(display_name):
...
+def _coerce_contrast_limits(contrast_limits: Tuple[float, float]):
+    """Coerce contrast limits to be in the float32 range."""
+    vispy_max = float(np.finfo(np.float32).max / 8)
+    if np.abs(contrast_limits).max() > vispy_max:
+        return scale_down(contrast_limits)
+
+    c_min = np.float32(contrast_limits[0])
+    if (
+        np.float32(contrast_limits[1]) - c_min
+        < (np.nextafter(c_min, np.float32(np.inf)) - c_min) * 64
```
Comment (psobolewskiPhD) [inline napari/utils/colormaps/colormap_utils.py]:
```suggestion
        < (np.nextafter(c_min, np.float32(np.inf)) - c_min) * 64
```
Can you add comment what the 64 does?

File: napari/utils/colormaps/colormap_utils.py [hunk 5]
```diff
@@ -906,3 +906,52 @@ def display_name_to_name(display_name):
...
+    return CoercedContrastLimits(contrast_limits, 0, 1)
+
+
+def scale_down(contrast_limits: Tuple[float, float]):
+    """Scale down contrast limits to be in the float32 range."""
+    vispy_max = float(np.finfo(np.float32).max / 8)
+    # Using 8 as divisor comes from experiments.
+    # For some reason if use smaller number,
+    # the image is not displayed correctly.
+    vispy_min = float(np.finfo(np.float32).min / 8)
```
Comment (psobolewskiPhD) [inline napari/utils/colormaps/colormap_utils.py]: This computes to `-vispy_max`

File: napari/utils/colormaps/_tests/test_colormap_utils.py [hunk 6]
```diff
@@ -38,3 +42,52 @@ def test_label_colormap_exception():
...
+    contrast_limits = (0.0, 1.0)
+    result = _coerce_contrast_limits(contrast_limits)
+    assert isinstance(result, CoercedContrastLimits)
+    assert np.allclose(result.contrast_limits, contrast_limits)
+    assert result.offset == 0
+    assert np.isclose(result.scale, 1.0)
+
+
+def test_coerce_contrast_limits_with_large_values():
+    contrast_limits = (0, 1e40)
```
Comment (psobolewskiPhD) [inline napari/utils/colormaps/_tests/test_colormap_utils.py]: Could we add a comment that 1e40 is larger than float32?

File: napari/utils/colormaps/_tests/test_colormap_utils.py [hunk 7]
```diff
@@ -38,3 +42,52 @@ def test_label_colormap_exception():
...
+    result = _coerce_contrast_limits(contrast_limits)
+    assert isinstance(result, CoercedContrastLimits)
+    assert np.isclose(result.contrast_limits[0], np.finfo(np.float32).min / 8)
+    assert np.isclose(result.contrast_limits[1], np.finfo(np.float32).max / 8)
+    assert result.offset < 0
+    assert result.scale < 1.0
+
+
+def test_coerce_contrast_limits_with_large_values_symetric():
```
Comment (psobolewskiPhD) [inline napari/utils/colormaps/_tests/test_colormap_utils.py]:
```suggestion
def test_coerce_contrast_limits_with_large_values_symmetric():
```

File: napari/utils/colormaps/_tests/test_colormap_utils.py [hunk 8]
```diff
@@ -38,3 +42,52 @@ def test_label_colormap_exception():
...
+    contrast_limits = (-1e40, 1e40)
```
Comment (psobolewskiPhD) [inline napari/utils/colormaps/_tests/test_colormap_utils.py]: Again, maybe worth a comment that both of these exceed float32

File: napari/layers/image/image.py [hunk 9]
```diff
@@ -534,6 +534,7 @@ def _update_slice_response(self, response: _ImageSliceResponse) -> None:
         """Update the slice output state currently on the layer. Currently used
         for both sync and async slicing.
         """
+        response = response.to_displayed(self._raw_to_displayed)
```
Comment (psobolewskiPhD) [inline napari/layers/image/image.py:537]: Maybe worth a comment that this is needed in analogous fashion to labels?
And update docstring of `to_displayed` in`_slice/_ImageSliceResponse`

File: napari/layers/image/image.py [hunk 10]
```diff
@@ -1191,3 +1192,40 @@ def _calc_data_range(self, mode='data') -> Tuple[float, float]:
...
+        ----------
+        raw : array
+            Raw array.
+
+        Returns
+        -------
+        image : array
+            Displayed array.
+        """
+        lim_tup = _coerce_contrast_limits(self.contrast_limits)
```
Comment (psobolewskiPhD) [inline napari/layers/image/image.py]: Could we call `lim_tup` something that more clearly indicates that it contains not just the limits tuple but also the other parameters (scale, offset)?
Maybe `coerced_params` or `coerced_limits_info`

Comment (psobolewskiPhD) [reply napari/layers/image/image.py:1211]: I mean it's sort of obvious that lim_tup == limits_tuple, so maybe ignore.

Comment (psobolewskiPhD) [reply napari/layers/image/image.py]: Or I guess could unpack the tuple into the limits tuple and the other two values.

Comment (psobolewskiPhD) [reply napari/layers/image/image.py]: Or how horrible is just using:
`_coerce_contrast_limits(self.contrast_limits).contrast_limits`?
and then
`_coerce_contrast_limits(self.contrast_limits).scale`

**Review by psobolewskiPhD — CHANGES_REQUESTED** [review]
Grzegorz walked me through this, it works and makes sense as explained.
Based on our conversation I left a few comments to make things clearer for the next person.

**Comment by Czaki** [comment]
@psobolewskiPhD I have addressed review.

@jnahlers could you check the current logic if works with your data?

File: napari/layers/image/image.py [hunk 11]
```diff
@@ -534,6 +534,10 @@ def _update_slice_response(self, response: _ImageSliceResponse) -> None:
         """Update the slice output state currently on the layer. Currently used
         for both sync and async slicing.
         """
+        response = response.to_displayed(self._raw_to_displayed)
+        # We call to_displayed here to ensure that if the contrast limits
+        # are outside the range of sup[ported by vispy, then data vie is
```
Comment (psobolewskiPhD) [inline napari/layers/image/image.py]:
```suggestion
        # are outside the range of supported by vispy, then data view is
```

**Review by psobolewskiPhD — APPROVED** [review]

**Comment by psobolewskiPhD** [comment]
@Czaki I think that test fail is real related to the uninitialized float64 `empty` array. Not sure if switching those arrays to zeroes would just mask something?

**Comment by Czaki** [comment]
> @Czaki I think that test fail is real related to the uninitialized float64 `empty` array. Not sure if switching those arrays to zeroes would just mask something?

No. The error was the result of mistaking scale and offset. For testability, I move logic to `CoercedContrastLimits`.

**Comment by jnahlers** [comment]
@Czaki I am still having trouble with the examples I [gave](https://github.com/napari/napari/pull/6537#issuecomment-1859281941) [above ](https://github.com/napari/napari/pull/6537#issuecomment-1859310100). Is this the issue of small contrast limits for big values you mentioned above?

**Comment by Czaki** [comment]
@jnahlers It looks like I have made a bug when merging main into this PR. I have fixed it now.

**Comment by Czaki** [comment]
As far as I could check, the tests are failing now because of `NaN`. So it looks like using `np.empty on float` is a bad idea. So it should be either a predefined array or int data. 

Random is bad because its slowdown test suite. 

cc @psobolewskiPhD

**Comment by psobolewskiPhD** [comment]
Even with int could we get something strange, given that fact that it's `uninitialized`?
Maybe best to use zeros?

Going forward, could we make a set of arrays of different size/shapes using random.random and then reuse them everywhere else?

**Comment by Czaki** [comment]
No. Int cannot be something strange.

We could create preset of data using fixtures.

**Comment by jnahlers** [comment]
@Czaki Ping me if you want to me try the branch again, right now I'm still seeing black on the examples. Thanks for you work on this!

**Comment by Czaki** [comment]
> @Czaki Ping me if you want to me try the branch again, right now I'm still seeing black on the examples. Thanks for you work on this!

I have removed an important change when removing debug print....

**Comment by Czaki** [comment]
This is how it looks on my machine:

https://github.com/napari/napari/assets/3826210/dd8f9aba-e6db-42fc-8291-568bb39ccc30

I do not understand the source of vertical artifacts. But as it looks like thumbnail do not have such artifacts, maybe rescale up could be done smarter.

**Comment by Czaki** [comment]
@jnahlers Is it ok now?

**Comment by jnahlers** [comment]
> @jnahlers Is it ok now?

Looks good!

**Pull request merged**